### PR TITLE
Position reconstruction neural network - first commit.

### PR DESCRIPTION
**Pull request by a1exndr**

### CODE REVIEW

File: straxbra/plugins.py [hunk 1]
```diff
@@ -321,68 +321,30 @@ class PeakPositions(strax.Plugin):
...
-        self.fitparameters = np.array([[ 0.58534179, 29.89341846, -0.3275816, 4.14715081, 2.61234684], 
-                          [ 0.63546614, 28.49525342, -0.19651583, 3.76011493, 2.68672152], 
-                          [ 0.58840586, 30.39015033, -0.38248759, 4.27642323, 2.56774842], 
-                          [ 0.59111988, 31.38350968, -0.5760228, 4.64122418, 2.51311592], 
-                          [ 0.63771524, 28.90663204, -0.26194541, 3.90052756, 2.66510948], 
-                          [ 0.59030322, 30.55082687, -0.46183924, 4.39769959, 2.55446814], 
-                          [ 0.53114467, 39.20861977, -17.93187819, 20.60397171, 2.27692367]])
-        self.fitparameters = self.fitparameters[self.pmt_mask[self.config['top_pmts']]]
+        ## Load the trained model from corresponding HDF5 file
+        self.model_NN = keras.models.load_model('XeBRA_Position_Reconstruction_NN_Model.h5')
```
Comment (darrylmasson) [inline straxbra/plugins.py]: Is this file in the folder where the script is running? Is it in the folder where the script is installed? Add the full path. Also, perhaps consider adding something to the filename indicating that it's for the dual-phase 7-PMT configuration.

File: straxbra/plugins.py [hunk 2]
```diff
@@ -321,68 +321,30 @@ class PeakPositions(strax.Plugin):
...
     def compute(self, peaks):
         # Keep large peaks only
         peak_mask = peaks['area'] > self.config['min_reconstruction_area']
         p = peaks['area_per_channel'][peak_mask, :]
-        p = p[:, self.pmt_mask]
         results = np.full_like(peaks, np.nan, dtype=self.dtype)
 
         for p_i,p in enumerate(peaks):
-            apc = p['area_per_channel'][self.pmt_mask]
+            apc = p['area_per_channel']
```
Comment (darrylmasson) [inline straxbra/plugins.py]: If the pmt mask isn't used, does this mean the area in the bottom channels are also part of the NN? If so, there isn't any need to have the top PMT list in the plugin config.

**Review by darrylmasson — CHANGES_REQUESTED** [review]
Looks good, only minor things to address

File: straxbra/plugins.py (hunk 2 above)
Comment (a1exndr) [reply straxbra/plugins.py]: Only the top channels are used, however the previous pmt mask excluding dead channels cannot be used since the NN requires an input shape of 7. The exclusion of the bottom PMT is now updated.

File: straxbra/plugins.py (hunk 1 above)
Comment (a1exndr) [reply straxbra/plugins.py]: The indication of the dual-phase 7-PMT configuration is added. The HDF5 file is currently in the same folder as plugins.py, therefore no path indication. What full path would you recommend to add resp. where is the file most conveniently stored?

Comment (darrylmasson) [reply straxbra/plugins.py]: Pick a directory (ideally on hydra somewhere, where no one will have issues accessing it).

File: straxbra/plugins.py [hunk 3]
```diff
@@ -321,29 +321,14 @@ class PeakPositions(strax.Plugin):
...
+        self.pmt_mask[self.config['top_pmts']] = np.ones_like(self.pmt_mask[self.config['top_pmts']])
+        ## Load the trained model from corresponding HDF5 file
+        self.model_NN = keras.models.load_model('XeBRA_Position_Reconstruction_NN_Model_DualPhase_7TopPMTs.h5')
         
     def compute(self, peaks):
-        # Keep large peaks only
+        ## Keep large peaks only
         peak_mask = peaks['area'] > self.config['min_reconstruction_area']
         p = peaks['area_per_channel'][peak_mask, :]
         p = p[:, self.pmt_mask]
```
Comment (darrylmasson) [inline straxbra/plugins.py]: What exactly do L332-334 accomplish? I don't see it

File: straxbra/plugins.py [hunk 4]
```diff
@@ -356,33 +341,14 @@ def compute(self, peaks):
...
         HFs_input = input_array / np.sum(input_array)
-        reconstruct = lambda x: np.sum(((self.LRF_PMTs(x[0], x[1]) - HFs_input)**2) / (self.LRF_PMTs(x[0], x[1])))
-        x0 = [0.001,0.001]
-        meth = 'SLSQP'
-        cons = ({'type': 'ineq', "fun": self.insidevolume})
-        res = minimize(reconstruct, x0, method=meth, constraints=cons)
-        return res.x
+        ## Use model to reconstruct position
+        ## Important: Factor 70 for rescaling label
+        predictions = self.model_NN.predict(np.array([HFs_input]))[0]*70
```
Comment (darrylmasson) [inline straxbra/plugins.py:347]: This '70' appears to be a magic number. I don't like it.

**Review by darrylmasson — CHANGES_REQUESTED** [review]
More minor changes

File: straxbra/plugins.py (hunk 1 above)
Comment (a1exndr) [reply straxbra/plugins.py]: Since most of the directories appear to have restricted access, I moved it to `/data/workspace/nn_models`.

File: straxbra/plugins.py (hunk 3 above)
Comment (a1exndr) [reply straxbra/plugins.py]: Seems to be a relic from the old code which is redundant and not used anymore. Removed it.

File: straxbra/plugins.py (hunk 4 above)
Comment (a1exndr) [reply straxbra/plugins.py:347]: The labels the NN was trained on (i.e. the x and y positions) were scaled to [-0.5, 0.5] in order to improve the efficiency and accuracy, so using this trained model we have to rescale to the physical positions.

Comment (darrylmasson) [reply straxbra/plugins.py:347]: For future work it may be better to scale the training data to [-1,1] because then your scaling factor is the TPC radius, which we would specify as an option.

File: straxbra/plugins.py [hunk 5]
```diff
@@ -321,32 +321,14 @@ class PeakPositions(strax.Plugin):
...
-                          [ 0.63546614, 28.49525342, -0.19651583, 3.76011493, 2.68672152], 
-                          [ 0.58840586, 30.39015033, -0.38248759, 4.27642323, 2.56774842], 
-                          [ 0.59111988, 31.38350968, -0.5760228, 4.64122418, 2.51311592], 
-                          [ 0.63771524, 28.90663204, -0.26194541, 3.90052756, 2.66510948], 
-                          [ 0.59030322, 30.55082687, -0.46183924, 4.39769959, 2.55446814], 
-                          [ 0.53114467, 39.20861977, -17.93187819, 20.60397171, 2.27692367]])
-        self.fitparameters = self.fitparameters[self.pmt_mask[self.config['top_pmts']]]
+        self.pmt_mask[self.config['top_pmts']] = np.ones_like(self.pmt_mask[self.config['top_pmts']])
+        ## Load the trained model from corresponding HDF5 file
+        self.model_NN = keras.models.load_model('/data/workspace/nn_models/XeBRA_Position_Reconstruction_NN_Model_DualPhase_7TopPMTs.h5')
```
Comment (darrylmasson) [inline straxbra/plugins.py:328]: At some point we should do some cleanup and move either the model (or at least the path to the model) into the database.

**Review by darrylmasson — APPROVED** [review]
Good enough to merge

File: straxbra/plugins.py (hunk 4 above)
Comment (a1exndr) [reply straxbra/plugins.py:347]: In that sense the current scaling factor is actually the TPC diameter, so if you want it's possible to specify it as an option.

File: straxbra/plugins.py (hunk 5 above)
Comment (a1exndr) [reply straxbra/plugins.py:328]: Definitely! This is just a temporary solution.
